### PR TITLE
Updated Data benchmarks

### PR DESCRIPTION
**Pull request by parkera**
This is a set of benchmarks for `struct Data`, slightly updated from https://github.com/apple/swift/pull/13997.

### CODE REVIEW

**Comment by parkera** [comment]
@swift-ci please benchmark

**Review by phausler — APPROVED** [review]
Looks good to me

**Comment by parkera** [comment]
@swift-ci please smoke test

**Comment by parkera** [comment]
@eeckstein what do you think about these?

File: benchmark/single-source/DataBenchmarks.swift [hunk 1]
```diff
@@ -0,0 +1,393 @@
...
+    case .veryLarge: return sampleData(size: 1024 * 1024 * 1024 + 128)
+    case .string: return sampleString()
+    case .immutableBacking: return sampleBridgedNSData()
+    }
+    
+}
+
+func benchmark_Subscript(_ N: Int, _ data: Data, _ index: Data.Index) {
+    for _ in 1...N {
+        _ = data[index]
```
Comment (eeckstein) [inline benchmark/single-source/DataBenchmarks.swift]: You should use blackHole(data[index]).
Just to make sure that the compiler does not optimize away the whole loop.
Data is a resilient type, so this should not happen. Just to be future prove.

File: benchmark/single-source/DataBenchmarks.swift [hunk 2]
```diff
@@ -0,0 +1,393 @@
...
+    }
+}
+
+func benchmark_Count(_ N: Int, _ data: Data) {
+    for _ in 1...10000*N {
+        _ = data.count
```
Comment (eeckstein) [inline benchmark/single-source/DataBenchmarks.swift]: Same here: blackHole(data.count)

File: benchmark/single-source/DataBenchmarks.swift [hunk 3]
```diff
@@ -0,0 +1,393 @@
...
+    for _ in 1...10000*N {
+        data.count = count
+        data.count = orig
+    }
+}
+
+func benchmark_AccessBytes(_ N: Int, _ data: Data) {
+    for _ in 1...10000*N {
+        data.withUnsafeBytes { (ptr: UnsafePointer<UInt8>) in
+            // do nothing here
```
Comment (eeckstein) [inline benchmark/single-source/DataBenchmarks.swift]: You should actually to something with the data, e.g. blackHole(ptr.pointee)

File: benchmark/single-source/DataBenchmarks.swift [hunk 4]
```diff
@@ -0,0 +1,393 @@
...
+    case .large: return sampleData(size: 40980)
+    case .veryLarge: return sampleData(size: 1024 * 1024 * 1024 + 128)
+    case .string: return sampleString()
+    case .immutableBacking: return sampleBridgedNSData()
+    }
+    
+}
+
+func benchmark_Subscript(_ N: Int, _ data: Data, _ index: Data.Index) {
+    for _ in 1...N {
```
Comment (eeckstein) [inline benchmark/single-source/DataBenchmarks.swift]: should this be 10000*N?

Also: you use 10000*N for all benchmarks. But I assume the time for one iteration varies a lot, e.g. for benchmark_Count vs benchmark_copyBytes.
Did you check (by running it locally) that the reported time is within a reasonable range for all benchmarks? That means > 10, ideally > 100.

File: benchmark/single-source/DataBenchmarks.swift [hunk 5]
```diff
@@ -0,0 +1,393 @@
...
+    case .medium: return sampleData(size: 1033)
+    case .large: return sampleData(size: 40980)
+    case .veryLarge: return sampleData(size: 1024 * 1024 * 1024 + 128)
+    case .string: return sampleString()
+    case .immutableBacking: return sampleBridgedNSData()
+    }
+    
+}
+
+func benchmark_Subscript(_ N: Int, _ data: Data, _ index: Data.Index) {
```
Comment (eeckstein) [inline benchmark/single-source/DataBenchmarks.swift]: Did you mean that those benchmark body functions really end up as separate functions? If so, you might add an @inline(never).

**Review by eeckstein — COMMENTED** [review]
Basically lgtm.

My only concern is that many benchmarks are pretty small. We have seen that such benchmarks are often noisy.
But we can let it run for a while and see how it looks like.

File: benchmark/single-source/DataBenchmarks.swift (hunk 1 above)
Comment (parkera) [reply benchmark/single-source/DataBenchmarks.swift]: Good idea, I'll do that.

File: benchmark/single-source/DataBenchmarks.swift (hunk 4 above)
Comment (parkera) [reply benchmark/single-source/DataBenchmarks.swift]: One iteration gives a value of 0, so yah I had to scale it up.

File: benchmark/single-source/DataBenchmarks.swift (hunk 5 above)
Comment (parkera) [reply benchmark/single-source/DataBenchmarks.swift]: I'm just going to simplify those out -- they were just following the pattern of the others that have a data argument (so we can call it with different sized data).

File: benchmark/single-source/DataBenchmarks.swift (hunk 4 above)
Comment (parkera) [reply benchmark/single-source/DataBenchmarks.swift]: Here is what I get locally with the suggested changes:

```
#,TEST,SAMPLES,MIN(us),MAX(us),MEAN(us),SD(us),MEDIAN(us)
1,DataAccessBytes,15,1276,1397,1328,42,1334
2,DataAppendArray,15,6051,6350,6237,95,6268
3,DataAppendBytes,15,5929,6348,6082,110,6072
4,DataAppendDataLargeToLarge,15,59606,77069,65998,5253,65009
5,DataAppendDataLargeToMedium,15,34978,37283,35716,739,35404
6,DataAppendDataLargeToSmall,15,33598,40655,35087,1628,34808
7,DataAppendDataMediumToLarge,15,22576,34445,24782,2934,23565
8,DataAppendDataMediumToMedium,15,8311,8848,8517,151,8480
9,DataAppendDataMediumToSmall,15,7897,8755,8174,200,8149
10,DataAppendDataSmallToLarge,15,22324,33598,24158,3589,22807
11,DataAppendDataSmallToMedium,15,7990,8534,8230,165,8223
12,DataAppendDataSmallToSmall,15,7574,8004,7784,129,7757
13,DataAppendSequence,15,19224,20106,19506,282,19471
14,DataCopyBytes,15,3007,3229,3123,56,3122
15,DataCount,15,52,54,52,1,53
16,DataMutateBytes,15,4776,4950,4868,52,4862
17,DataReplaceLarge,15,40273,43275,41635,1009,41217
18,DataReplaceLargeBuffer,15,56077,59662,57944,878,57766
19,DataReplaceMedium,15,13324,13753,13549,123,13537
20,DataReplaceMediumBuffer,15,13211,13734,13461,130,13459
21,DataReplaceSmall,15,10941,11486,11143,132,11146
22,DataReplaceSmallBuffer,15,10484,11222,10690,181,10711
23,DataReset,15,3397,3787,3579,102,3585
24,DataSetCount,15,919,1027,975,33,974
25,DataSubscript,15,284,297,288,4,289
```

Comment (parkera) [reply benchmark/single-source/DataBenchmarks.swift]: Oh that's 15 samples.

**Comment by parkera** [comment]
@swift-ci smoke test

**Comment by parkera** [comment]
@swift-ci benchmark

**Comment by parkera** [comment]
@swift-ci please benchmark

**Review by eeckstein — APPROVED** [review]

**Comment by airspeedswift** [comment]
It still looks like the variance on some of these are above the threshold (it's 5%?) for reporting regressions. I really don't want to add more "always report regression" tests. Any chance we could split out the unstable ones and iterate on those a bit more to make them stable, while landing the rest?

**Comment by parkera** [comment]
Is it 5% or 10%?

**Comment by airspeedswift** [comment]
Looking at the report above, it looks like 5%

**Comment by parkera** [comment]
@eeckstein should I mark them as `.unstable` until we have more benchmarking data?

**Comment by eeckstein** [comment]
@parkera Yes, that would make sense

**Comment by palimondo** [comment]
Re: discussions about unstable tests, please have a look at the post [Towards Robust Performance Measurement](https://forums.swift.org/t/towards-robust-performance-measurement/11490) in the forum.

**Pull request merged**